### ISSUE TITLE
keyboard control is looping around the charts , not reading the contents of the chart.

### ISSUE THREAD

**Issue by Sandhya-aluri** [highcharts/highcharts#10615]
#### Product version
highcharts: ^7.0.3,

**Comment by sebastianbochan** [comment]
Hi @Sandhya-aluri,
Thank you for reporting about the problem.

FYI @oysteinmoseng

**Comment by oysteinmoseng** [comment]
Hi @Sandhya-aluri.

JAWS/NVDA are able to read the data values for me (example: https://www.highcharts.com/samples/highcharts/demo/accessible-line) on charts with the [accessibility module](https://www.highcharts.com/docs/accessibility/accessibility-module) enabled. In JAWS, for example, you can use down arrow to navigate the chart using the virtual cursor. If you are using tab, you need to use the arrow keys to move between data points. This is not easily possible in JAWS/NVDA, because the arrow keys are bound to different screen reader keyboard shortcuts, and Highcharts can not override this. If you still want to use arrow keys, you can press Insert+3 in JAWS to pass through the next key. This will allow you to use your arrow keys as if JAWS was not running.

The most efficient way to navigate the charts with JAWS/NVDA is with the virtual cursor, however.

I am closing this for now, feel free to reopen if this does not answer your question.

**Comment by matatk** [comment]
In general, virtual cursor navigation is helpful because it frees people using screen readers to navigate around the page in a number of different ways. However, in this case, it's most natural to navigate around the domain-specific objects in the chart—the data series and datum points—rather than trawl through the DOM with the virtual cursor (which causes the screen reader to announce all the SVG path objects, with no way to skip directly between datum points or series).

Being able to quickly move between series with the up/down arrow keys, or between datum points in a series with left/right, seems really helpful to all users—especially people using a screen reader, as it provides a time-saving shortcut.

@oysteinmoseng are there any particular reasons for the current behaviour, of having screen reader users use the virtual cursor by default?

The standard data-oriented Highcharts keyboard navigation can be restored for screen reader users by adding a `role="application"` attribute to the focusable chart container.

[ Specifically, @aardrian and I have been looking at the [accessible line plot demo](https://www.highcharts.com/samples/highcharts/accessibility/accessible-line/). ]

**Comment by pawelfus** [comment]
Hi @matatk - just FYI: @oysteinmoseng will be back next week

**Comment by oysteinmoseng** [comment]
Hi @matatk,

It is a very valid concern that the custom keyboard navigation features of Highcharts can have benefits for screen reader users as well, and we have been considering a `role="application"` solution for some time. In our user testing so far, however, screen reader users have not been positive towards a `role="application"` approach. I believe the reason for this is twofold.

Firstly, the keyboard navigation features are optimized to be intuitive when navigating from a visual perspective. Pressing the "up" arrow key moves to the nearest point vertically for most chart types. Concepts such as "left", "right", "up", and "down" are spatial in nature, and keeping track of the spatial relationships between data points can introduce a significant cognitive load. It can even be argued that this kind of navigation for a screen reader user is in violation of [WCAG 1.3.3](https://www.w3.org/WAI/WCAG21/Understanding/sensory-characteristics.html).

Now, it would of course be possible to create a simplified custom navigation model for screen reader users, which is something we have considered, but that brings us to the second issue: There is a learning curve associated with custom navigation - along with discoverability issues, further adding to the cognitive load. In my experience, most screen reader users are immediately skeptical towards `role="application"` encounters for this reason. By using standard semantics we are able to leverage the native capabilities of the user's assistive technology, and they are able to use controls and features that are familiar to them.

We have tried to help with the navigation challenges by allowing each data series to be marked up as a landmark (region), and by adding headings that can be used for skip navigation. These features are still being tweaked to find an intuitive set of defaults, but we are trying to avoid the `role="application"` approach for now, for the reasons above. Since our goal is to remove as many barriers as possible for users to experience our charts, this seems to be the best strategy given the current options. We are happy to learn about other perspectives, however, and welcome any further suggestions or feedback you might have.

**Comment by matatk** [comment]
Hi @oysteinmoseng, thanks for your thoughtful reply. We have some further thoughts and questions, which we hope may be of help.
 
First of all, we understand that you, as the developers of Highcharts, need to consider all chart types (such as maps), whereas currently we're specifically looking at line/bar/pie charts, which have a more logical arrow-key focus order (which effectively mirrors the underlying data table).
 
We also understand the reticence to use the `application` role, as it has been misused in the past, though in this case, it does seem like an appropriate way to provide keyboard navigation.
 
Highcharts presentation of the structure using landmarks is well thought-out and I'm sure would suit the way that a significant number of screen-reader users would like to navigate. However, there are some thoughts and observations we had about it...
 
1. It's non-visual, and a significant number of screen-reader users (between a quarter and a third, according to a recent WebAIM surveys [0,1]) have some sight, and use the screen-reader in conjunction with it. Whilst it is very logical and may be really helpful for people who are blind, it could be very confusing for people who are looking and can't see the corresponding structure on the screen, but can hear them being announced.
 
2. It's possible to use to the virtual cursor means of navigation even if there is `role="application"` on the main focusable chart container. This could be done either by: navigating into the chart using the virtual cursor or manually disabling it through the use of a shortcut key provided by the screen-reader. Perhaps you could implement a toggle for screen-reader users to help them switch between the modes? (Some research would need to be done to ensure this is feasible, but SRs have reacted well to our manual adding/removal of the role so far.)
 
3. The landmarks-based method doesn't allow the user to skip directly between a point in one series and its corresponding point in another, but the arrow-keys method does (in charts that have a tabular logical structure, at least, as mentioned above).
 
In the case of charts such as maps (e.g. the one from #10615), it seems that the arrow key focus order may be more relatable to all users if it followed the 2D layout (as much as possible) rather than the progression along the data series, but that is definitely a harder problem to resolve.
 
@matatk, @aardrian
 
[0] https://webaim.org/projects/screenreadersurvey7/#disabilitytypes
[1] https://webaim.org/projects/screenreadersurvey6/#disabilitytypes

**Comment by oysteinmoseng** [comment]
@matatk @aardrian Great input, and I appreciate the time and effort you put into this.

I agree that the landmark navigation is far from as good as the keyboard navigation for visual or semi-visual users. The challenge is finding an intuitive way to provide a screen reader user (that may or may not have some vision) with a complete navigation solution. Since we have no way of knowing whether or not the screen reader user will prefer virtual cursor navigation or a more visual approach, we are probably looking at a toggle switch of some kind, as you suggest. We have visited this option before, but need to find a good, intuitive design. I think it is definitely worth experimenting with.

We are also looking at navigation issues when it comes to chart sonification for screen reader users, and it is possible that some of these challenges overlap. We also have some interesting challenges with tree-structured charts (network graphs, treemaps etc). These will almost certainly require custom navigation to be usable.

Ultimately, I think this is one of those areas where we need to come up with some potential designs and then do extensive user testing.
